zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java

```diff
@@ -238,6 +239,9 @@ public void run() {
   @Override
   public void shutdown() throws TException {
     logger.info("Shutting down...");
+    // delete interpreter cluster meta
+    deleteClusterMeta();
```

Comment:
What happens if interpreter process is killed manually (e.g. via kill -9, or killed by yarn) in which case shutdown is not called.

Reply:
To determine if an interpreter is available, there are two ways to cluster.
1. The interpreter process exits normally, where the interpreter actively deletes the information in the cluster metadata.
2. The interpreter exits abnormally. zeppelin-server periodically checks if the interpreter is available. If it is not available, it will be set to offline.
https://github.com/apache/zeppelin/blob/master/zeppelin-interpreter/src/main/java/org/apache/zeppelin/cluster/ClusterMonitor.java#L165